### ISSUE TITLE
Lint idea: async functions marked to return FutureOr<T>

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/sdk#58022]
There was some confusion on a flutter chat amongst some pretty dang smart/experienced Dart devs around whether an async function could return a non-future value, or if there was any way to return a non-future value and still use await.

Some of the confusion is around how its legal to write:

```
FutureOr<int> asyncIntFunction(FutureOr<int> input) async {
  if (input is int) {
    return 0;
  } else {
    await input;
    return Future.value(0);
  }
}
```

This sure *reads* like it's returning a `Future<int>` when the input is `Future<int>`, and `int` when the input is `int`. However, this is _not_ what's happening.

Any async function's return value is always wrapped in a Future unless it already is one. So the function `asyncIntFunction` will never return an `int`. And the return type may as well have been written as `Future<int>`.

Writing a return type of `FutureOr<int>` is legal just like writing a return type of `dynamic` or `Object` or `Future<dynamic>` is legal: it's always OK to return a wider type statically. (The spec says _"It is a compile-time error if the declared return type of a function marked `async` is not a supertype of `Future<T>` for some type `T`."_

However in this case it really just obscures some of the steps that you have to actually go through to actually get the behavior you want, which is:

```
FutureOr<int> asyncIntFunction(FutureOr<int> input) {
  if (input is int) {
    return 0;
  } else {
    return input.then((_) => 0);
  }
}
```

So, in summary, writing `FutureOr<T> f async ...` probably is an indication that you are in the act of misunderstanding how `async` and `FutureOr` actually work together, and giving no error/hint/lint here is giving the user a footgun.
